src/hippopt/robot_planning/variables/contacts.py

```diff
@@ -0,0 +1,169 @@
+import dataclasses
+from typing import TypeVar
+
+import casadi as cs
+import liecasadi
+import numpy as np
+
+from hippopt import (
+    CompositeType,
+    OptimizationObject,
+    OverridableVariable,
+    Parameter,
+    StorageType,
+    default_composite_field,
+    default_storage_field,
+)
+
+
+@dataclasses.dataclass
+class ContactPointDescriptor(OptimizationObject):
+    position_in_foot_frame: StorageType = default_storage_field(Parameter)
+    foot_frame: str = dataclasses.field(default=None)
+
+    input_foot_frame: dataclasses.InitVar[str] = dataclasses.field(default=None)
+    input_position_in_foot_frame: dataclasses.InitVar[np.ndarray] = dataclasses.field(
+        default=None
+    )
+
+    def __post_init__(
+        self, input_foot_frame: str, input_position_in_foot_frame: np.ndarray
+    ) -> None:
+        if input_foot_frame is not None:
+            self.foot_frame = input_foot_frame
+        if input_position_in_foot_frame is not None:
+            self.position_in_foot_frame = input_position_in_foot_frame
+
+    @staticmethod
+    def rectangular_foot(
+        foot_frame: str,
+        x_length: float,
+        y_length: float,
+        top_left_point_position: np.array,
+    ):
+        return [
+            ContactPointDescriptor(
+                input_foot_frame=foot_frame,
+                input_position_in_foot_frame=top_left_point_position,
+            ),
+            ContactPointDescriptor(
+                input_foot_frame=foot_frame,
+                input_position_in_foot_frame=top_left_point_position
+                + np.array([-x_length, 0.0, 0.0]),
+            ),
+            ContactPointDescriptor(
+                input_foot_frame=foot_frame,
+                input_position_in_foot_frame=top_left_point_position
+                + np.array([-x_length, -y_length, 0.0]),
+            ),
+            ContactPointDescriptor(
+                input_foot_frame=foot_frame,
+                input_position_in_foot_frame=top_left_point_position
+                + np.array([0.0, -y_length, 0.0]),
+            ),
+        ]
+
+
+@dataclasses.dataclass
+class ContactPointState(OptimizationObject):
+    p: StorageType = default_storage_field(OverridableVariable)
+    f: StorageType = default_storage_field(OverridableVariable)
+
+    descriptor: CompositeType[ContactPointDescriptor] = default_composite_field(
+        factory=ContactPointDescriptor, time_varying=False
+    )
+
+    input_descriptor: dataclasses.InitVar[ContactPointDescriptor] = dataclasses.field(
+        default=None
+    )
+
+    def __post_init__(self, input_descriptor: ContactPointDescriptor) -> None:
+        if self.p is None:
+            self.p = np.zeros(3)
```

Comment:
Pattern change. Used in many other places.

```suggestion
        self.p = np.zeros(3) if self.p is None else self.p
```

Reply:
Thanks! Done in https://github.com/ami-iit/hippopt/pull/10/commits/57e4b49d4ebe8b71b12a4dc36a3f9fdc9cbb7c69